dali/operators/util/get_property.cc

```diff
@@ -0,0 +1,162 @@
+    auto& output = ws.template Output<Backend>(0);
+    for (size_t sample_id = 0; sample_id < input.num_samples(); sample_id++) {
+      auto si = GetSourceInfo(input, sample_id);
+      output[sample_id].Copy(make_cspan((const uint8_t*)si.c_str(), si.length()), nullptr);
+    }
+  }
+
+ private:
+  const std::string& GetSourceInfo(const BatchContainer& input, size_t idx) {
+    return input[idx].GetMeta().GetSourceInfo();
+  }
+};
+
+template <>
+const std::string& SourceInfo<GPUBackend>::GetSourceInfo(const TensorList<GPUBackend>& input,
+                                                         size_t idx) {
+  return input.GetMeta(static_cast<int>(idx)).GetSourceInfo();
+}
+
+template <>
+void SourceInfo<GPUBackend>::FillOutput(workspace_t<GPUBackend>& ws) {
+  const auto& input = ws.template Input<GPUBackend>(0);
+  auto& output = ws.template Output<GPUBackend>(0);
+  for (size_t sample_id = 0; sample_id < input.num_samples(); sample_id++) {
+    auto si = GetSourceInfo(input, sample_id);
+    auto output_ptr = output.raw_mutable_tensor(static_cast<int>(sample_id));
+    assert(ws.has_stream());  // GPU operator must have CUDA stream assigned
+    cudaMemcpyAsync(output_ptr, si.c_str(), si.length(), cudaMemcpyDefault, ws.stream());
+  }
+}
+
+
+template <typename Backend, typename BatchContainer = batch_container_t<Backend>>
+struct Layout : public Property<Backend> {
+  TensorListShape<> GetShape(const BatchContainer& input) override {
+    // Every tensor in the output has the same number of dimensions
+    return uniform_list_shape(input.num_samples(), {GetLayout(input, 0).size()});
+  }
+
+  DALIDataType GetType(const BatchContainer&) override {
+    return DALI_UINT8;
+  }
+
+  void FillOutput(workspace_t<Backend>& ws) override {
+    const auto& input = ws.template Input<Backend>(0);
+    auto& output = ws.template Output<Backend>(0);
+    for (size_t sample_id = 0; sample_id < input.num_samples(); sample_id++) {
+      auto layout = GetLayout(input, sample_id);
+      output[sample_id].Copy(make_cspan((const uint8_t*)layout.c_str(), layout.size()), nullptr);
+    }
+  }
+
+ private:
+  const TensorLayout& GetLayout(const BatchContainer& input, int idx) {
+    return input[idx].GetMeta().GetLayout();
+  }
+};
+
+template <>
+const TensorLayout& Layout<GPUBackend>::GetLayout(const TensorList<GPUBackend>& input, int idx) {
+  return input.GetMeta(idx).GetLayout();
+}
+
+template <>
+void Layout<GPUBackend>::FillOutput(workspace_t<GPUBackend>& ws) {
+  const auto& input = ws.template Input<GPUBackend>(0);
+  auto& output = ws.template Output<GPUBackend>(0);
+  for (size_t sample_id = 0; sample_id < input.num_samples(); sample_id++) {
+    auto layout = GetLayout(input, sample_id);
+    auto output_ptr = output.raw_mutable_tensor(static_cast<int>(sample_id));
+    assert(ws.has_stream());  // GPU operator must have CUDA stream assigned
+    cudaMemcpyAsync(output_ptr, layout.c_str(), layout.size(), cudaMemcpyDefault, ws.stream());
```

Comment:
Copy does this for you, no need to specialize for GPU I think.

Reply:
I'm not sure, which `Copy` you are referring to. I guess it is about `Tensor::Copy` function, which I used for the CPU variant. The `input` type in the GPU variant is a `TensorList` and it does not have such `Copy` function which would allow accessing tensors in a batch one by one. I'd gladly see some generalization of `FillOutput` function for both backend, but I'm afraid it is not possible. I mean, it is possible to some extent, but maybe it's not worth obfuscating the code just to save 3 LOC.

Reply:
OK